content/Training/developers/05-guardrails-evals-observability.mdx

```diff
@@ -0,0 +1,1114 @@
+        # Add context about this operation
+        span.set_attribute("user.tier", "premium")
+        span.set_attribute("query.type", "retirement")
+        span.set_attribute("query.length", len(query))
+        
+        try:
+            # Track validation steps
+            span.add_event("validation-started")
+            
+            if has_prohibited_terms(query):
+                span.add_event("validation-failed", {"reason": "prohibited-terms"})
+                span.set_status(Status(StatusCode.ERROR, "Prohibited terms detected"))
+                return response.json({"error": "Invalid query"})
+            
+            span.add_event("validation-passed")
+            
+            # Track LLM call separately
+            async with context.tracer.start_as_current_span("generate-advice") as llm_span:
+                llm_span.set_attribute("model", "gpt-4")
+                advice = await generate_advice(query)
+                llm_span.set_attribute("response.tokens", count_tokens(advice))
+            
+            span.set_status(Status(StatusCode.OK))
+            return response.json({"advice": advice})
+            
+        except Exception as e:
+            span.record_exception(e)
+            span.set_status(Status(StatusCode.ERROR, str(e)))
+            raise`} js={`import { SpanStatusCode } from '@opentelemetry/api';
+
+const handler = async (request, response, context) => {
+  // Create a span for the entire validation flow
+  return context.tracer.startActiveSpan('validate-financial-query', async (span) => {
+    // Add context about this operation
+    span.setAttribute('user.tier', 'premium');
+    span.setAttribute('query.type', 'retirement');
+    span.setAttribute('query.length', query.length);
+    
+    try {
+      // Track validation steps
+      span.addEvent('validation-started');
+      
+      if (hasProhibitedTerms(query)) {
+        span.addEvent('validation-failed', { reason: 'prohibited-terms' });
+        span.setStatus({ code: SpanStatusCode.ERROR, message: 'Prohibited terms detected' });
+        return response.json({ error: 'Invalid query' });
+      }
+      
+      span.addEvent('validation-passed');
+      
+      // Track LLM call separately
+      const advice = await context.tracer.startActiveSpan('generate-advice', async (llmSpan) => {
+        llmSpan.setAttribute('model', 'gpt-4');
+        const result = await generateAdvice(query);
+        llmSpan.setAttribute('response.tokens', countTokens(result));
+        llmSpan.end();
+        return result;
+      });
+      
+      span.setStatus({ code: SpanStatusCode.OK });
+      return response.json({ advice });
+      
+    } catch (error) {
+      span.recordException(error);
+      span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
+      throw error;
+    } finally {
+      span.end();
+    }
+  });
+};`} />
+
+### Performance Best Practices
```

Comment:
I kinda think this should be at the beginning of the course.  start with our main value prop of observability and then lean in to performance and such.